google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -128,9 +143,22 @@ public void onNext(StreamingPullResponse response) {
           new Runnable() {
             @Override
             public void run() {
-              // Only if not shutdown we will request one more batches of messages to be delivered.
+              /**
+               * Only if not shutdown we will request one more batches of messages to be delivered.
+               *
+               * <p>We use the request observer we're paired with, not the "current observer" in the
+               * outer class. By the time this Runnable is called, the stream might have already
+               * failed and the outer class might have created a new stream. In this case, we should
+               * request more messages from the already-failed stream (a no-op), otherwise the
+               * current observer might get more messages than it should.
+               */
               if (isAlive()) {
-                requestObserver.request(1);
+                lock.lock();
+                try {
+                  thisRequestObserver.request(1);
```

Comment:
This can also be done under a reader lock, the only writer lock we should need to hold is while we are aborting and restarting the stream.

Reply:
I don't think a read lock is right here. `StreamObserver`s don't need to be thread-safe. If we make this a read-lock, it is possible for us to call `request` and `onNext` concurrently.

Reply:
If you see ClientCall.java line 177, while StreamObserver is not necessarily thread-safe, the request method in particular is guaranteed to be able to be called concurrently without external synchronization.

Reply:
Ah I see. Makes sense. Done